concent_api/verifier/tasks.py

```diff
@@ -24,28 +35,95 @@ def blender_verification_order(
 
 
 @shared_task
+@transaction.atomic(using='control')
 def verification_result(
     subtask_id:     str,
     result:         VerificationResult,
-    error_message:  str,
-    error_code:     str,
+    error_message:  Optional[str] = None,
+    error_code:     Optional[ErrorCode]= None,
 ):
-    logger.info(
-        'verification_result_task starts with: SUBTASK_ID {} -- RESULT {}'.format(
-            subtask_id,
-            result,
-        )
-    )
+    logger.info(f'verification_result_task starts with: SUBTASK_ID {subtask_id} -- RESULT {result}')
 
     assert isinstance(subtask_id, str)
-    assert isinstance(error_message, str)
-    assert isinstance(error_code, str)
-    assert result in VerificationResult
+    assert isinstance(result, VerificationResult)
+    assert isinstance(error_message, (str, type(None)))
+    assert isinstance(error_code, (ErrorCode, type(None)))
     assert all([error_message, error_code]) if result == VerificationResult.ERROR else True
 
-    logger.info(
-        'verification_result_task ends with: SUBTASK_ID {} -- RESULT {}'.format(
-            subtask_id,
-            result,
+    # Worker locks database row corresponding to the subtask in the subtask table.
+    try:
+        subtask = Subtask.objects.select_for_update(nowait=True).get(subtask_id=subtask_id)
+    except DatabaseError:
+        # If the row is already locked, task fails so that Celery can retry later.
```

Comment:
Just an idea: shouldn't we limit the number of retries and/or log the attempts?

Reply:
Definitely log it.

As for limiting retries - yeah, we should. It's not stated explicitly in the issue but having a way out of an infinite loop is a good idea :) But don't set the limit too low. It's much, much more likely that some other process is just holding the lock.

Reply:
How would you like to store it ? The only way I see is to add a field on `Subtask`. Is that ok ?

Reply:
another argument passed to `verification_result` (the number of attempt), maybe with default value 0?

Reply:
Good idea, lets go this way.

Reply:
Celery already stores the number of retries:

[Task.max_retries](http://docs.celeryproject.org/en/latest/reference/celery.app.task.html#celery.app.task.Task.max_retries)
[Task.retry()](http://docs.celeryproject.org/en/latest/reference/celery.app.task.html#celery.app.task.Task.retry)
[Task.apply()](http://docs.celeryproject.org/en/latest/reference/celery.app.task.html#celery.app.task.Task.apply) has a parameter called `retries`. This values is also accessible in some way when you get the task.

Reply:
`apply` is the blocking call, so we don't want it, but `retry` and `max_retries` looks good for this case and should be enough.

Reply:
There are also retry policies that let you e.g. use increasing intervals between retries. See [Message Sending Retry](http://docs.celeryproject.org/en/latest/userguide/calling.html#message-sending-retry)

Reply:
> `apply` is the blocking call, so we don't want it

I was just an example. `apply_async()` has it too.

Reply:
:+1: done